controllers/chat/sendAnonymousMessage.js

```diff
@@ -0,0 +1,101 @@
+const {MESSAGE_TYPE} = require('../../helpers/constants');
+const {determineMessageType, processReplyMessage} = require('../../utils/chat');
+
+const v = new Validator();
+const sendAnonymousMessage = async (req, res) => {
+  const schema = {
+    channelId: 'string|empty:false',
+    message: 'string|empty:false',
+    messageType: {
+      type: 'enum',
+      values: Object.values(MESSAGE_TYPE),
+      empty: true,
+      nullable: true,
+      optional: true
+    },
+    replyMessageId: 'string|empty:true|nullable:true|optional:true',
+    attachments: {
+      type: 'array',
+      optional: true,
+      nullable: true,
+      empty: true,
+      items: {
+        type: 'object',
+        props: {
+          type: 'string',
+          asset_url: 'string',
+          thumb_url: 'string',
+          myCustomField: 'string'
+        }
+      }
+    }
+  };
+  const validated = v.validate(req.body, schema);
+  if (validated.length)
+    return res.status(403).json({
+      message: 'Error validation',
+      error: validated
+    });
+
+  const {channelId, message, attachments, messageType, replyMessageId} = req.body;
+
+  const client = StreamChat.getInstance(process.env.API_KEY, process.env.SECRET);
+  try {
+    await client.connectUser({id: req.userId}, req.token);
+
+    const channel = client.channel('messaging', channelId);
+
+    const createdChannel = await channel.create();
+
+    if (
+      req.user.is_anonymous &&
+      createdChannel.channel.anon_user_info_emoji_name &&
+      client.user.name !== `Anonymous ${createdChannel.channel.anon_user_info_emoji_name}`
+    ) {
+      await client.upsertUser({
+        id: req.userId,
+        name: `Anonymous ${createdChannel.channel.anon_user_info_emoji_name}`,
+        image: createdChannel.channel.anon_user_info_emoji_code,
+        username: `Anonymous ${createdChannel.channel.anon_user_info_emoji_name}`
+      });
+    }
+    if (createdChannel?.channel?.is_channel_blocked)
+      return res.status(403).json(responseError('Channel is blocked'));
+
+    const currentMessageType = determineMessageType(messageType, attachments);
+    let baseMessage = {
+      user_id: req.userId,
+      attachments,
+      message_type: currentMessageType,
+      text: message
+    };
+
+    if (messageType === MESSAGE_TYPE.REPLY) {
+      baseMessage = await processReplyMessage({
+        res,
+        baseMessage,
+        channelId,
+        client,
+        messageType,
+        replyMessageId
+      });
+    }
+
+    const chat = await channel.sendMessage(baseMessage);
+
+    await client.disconnectUser();
+    return res.status(200).json(responseSuccess('sent', chat));
+  } catch (error) {
+    await client.disconnectUser();
+    return res.status(error.statusCode ?? error.status ?? 400).json({
+      status: 'error',
+      code: error.statusCode ?? error.status ?? 400,
+      message: error.message
+    });
+  }
+};
+
+module.exports = sendAnonymousMessage;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The code looks good overall, but there are a few areas that could be improved for better error handling and maintainability.

1. Error Handling: In the catch block (lines 91-98), you're assuming that `error` will always have `statusCode`, `status`, or `message` properties. This might not always be the case, especially if an unexpected error occurs. It would be safer to add a default error message and status code.

2. Maintainability: The validation schema (lines 9-34) is defined inside the function, which makes the function quite large and harder to read. Consider moving this schema to a separate file or at least outside of the function.

3. Security: You're directly using `process.env.API_KEY` and `process.env.SECRET` in line 45. It's generally a good practice to avoid direct usage of environment variables throughout your codebase. Instead, consider creating a config file where you import all your environment variables and then import them from there in your other files. This approach has two benefits: it centralizes the management of environment variables, and it allows you to validate these variables when your application starts.

Here are the suggested changes:

```diff
const Validator = require('fastest-validator');
const {StreamChat} = require('stream-chat');
const {responseError, responseSuccess} = require('../../utils/Responses');
const {MESSAGE_TYPE} = require('../../helpers/constants');
const {determineMessageType, processReplyMessage} = require('../../utils/chat');

const v = new Validator();
+ const config = require('../../config');

// Move schema outside of the function
+ const schema = {
+   channelId: 'string|empty:false',
+   message: 'string|empty:false',
+   messageType: {
+     type: 'enum',
+     values: Object.values(MESSAGE_TYPE),
+     empty: true,
+     nullable: true,
+     optional: true
+   },
+   replyMessageId: 'string|empty:true|nullable:true|optional:true',
+   attachments: {
+     type: 'array',
+     optional: true,
+     nullable: true,
+     empty: true,
+     items: {
+       type: 'object',
+       props: {
+         type: 'string',
+         asset_url: 'string',
+         thumb_url: 'string',
+         myCustomField: 'string'
+       }
+     }
+   }
+ };

const sendAnonymousMessage = async (req, res) => {
-  const schema = {
-    channelId: 'string|empty:false',
-    message: 'string|empty:false',
-    messageType: {
-      type: 'enum',
-      values: Object.values(MESSAGE_TYPE),
-      empty: true,
-      nullable: true,
-      optional: true
-    },
-    replyMessageId: 'string|empty:true|nullable:true|optional:true',
-    attachments: {
-      type: 'array',
-      optional: true,
-      nullable: true,
-      empty: true,
-      items: {
-        type: 'object',
-        props: {
-          type: 'string',
-          asset_url: 'string',
-          thumb_url: 'string',
-          myCustomField: 'string'
-        }
-      }
-    }
-  };
  const validated = v.validate(req.body, schema);
  if (validated.length)
    return res.status(403).json({
      message: 'Error validation',
      error: validated
    });

  const {channelId, message, attachments, messageType, replyMessageId} = req.body;

- 45:   const client = StreamChat.getInstance(process.env.API_KEY, process.env.SECRET);
+ 45:   const client = StreamChat.getInstance(config.API_KEY, config.SECRET);
  try {
    await client.connectUser({id: req.userId}, req.token);

    const channel = client.channel('messaging', channelId);

    const createdChannel = await channel.create();

    if (
      req.user.is_anonymous &&
      createdChannel.channel.anon_user_info_emoji_name &&
      client.user.name !== `Anonymous ${createdChannel.channel.anon_user_info_emoji_name}`
    ) {
      await client.upsertUser({
        id: req.userId,
        name: `Anonymous ${createdChannel.channel.anon_user_info_emoji_name}`,
        image: createdChannel.channel.anon_user_info_emoji_code,
        username: `Anonymous ${createdChannel.channel.anon_user_info_emoji_name}`
      });
    }
    if (createdChannel?.channel?.is_channel_blocked)
      return res.status(403).json(responseError('Channel is blocked'));

    const currentMessageType = determineMessageType(messageType, attachments);
    let baseMessage = {
      user_id: req.userId,
      attachments,
      message_type: currentMessageType,
      text: message
    };

    if (messageType === MESSAGE_TYPE.REPLY) {
      baseMessage = await processReplyMessage({
        res,
        baseMessage,
        channelId,
        client,
        messageType,
        replyMessageId
      });
    }

    const chat = await channel.sendMessage(baseMessage);

    await client.disconnectUser();
    return res.status(200).json(responseSuccess('sent', chat));
  } catch (error) {
- 92:     await client.disconnectUser();
- 93:     return res.status(error.statusCode ?? error.status ?? 400).json({
- 94:       status: 'error',
- 95:       code: error.statusCode ?? error.status ?? 400,
- 96:       message: error.message
- 97:     });
+ 92:     try {
+ 93:       await client.disconnectUser();
+ 94:     } catch (disconnectError) {
+ 95:       console.error('Failed to disconnect user:', disconnectError);
+ 96:     }
+ 97:     return res.status(error.statusCode ?? error.status ?? 500).json({
+ 98:       status: 'error',
+ 99:       code: error.statusCode ?? error.status ?? 500,
+ 100:      message: error.message || 'An unexpected error occurred'
+ 101:    });
  }
};

module.exports = sendAnonymousMessage;
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->